pandas-stubs/_libs/tslibs/period.pyi

```diff
@@ -1,34 +1,137 @@
-from typing import Any
+import datetime
+from typing import (
+    Literal,
+    Union,
+    overload,
+)
+
+import numpy as np
+from pandas import (
+    Index,
+    PeriodIndex,
+    Timedelta,
+    TimedeltaIndex,
+)
+from pandas.core.series import (
+    PeriodSeries,
+    TimedeltaSeries,
+)
+from typing_extensions import TypeAlias
+
+from pandas._libs.tslibs import NaTType
+from pandas._typing import npt
+
+from .timestamps import Timestamp
 
 class IncompatibleFrequency(ValueError): ...
 
-class Period:
+from pandas._libs.tslibs.offsets import BaseOffset
+
+_PeriodAddSub: TypeAlias = Union[
+    Timedelta, datetime.timedelta, np.timedelta64, np.int64, int, BaseOffset
+]
+
+_PeriodFreqHow: TypeAlias = Literal[
+    "S",
+    "E",
+    "start",
+    "end",
+]
+
+_PeriodToTimestampHow: TypeAlias = Union[
+    _PeriodFreqHow,
+    Literal[
+        "Start",
+        "Finish",
+        "Begin",
+        "End",
+        "s",
+        "e",
+        "finish",
+        "begin",
+    ],
+]
+
+class PeriodMixin:
+    @property
+    def end_time(self) -> Timestamp: ...
+    @property
+    def start_time(self) -> Timestamp: ...
+
+class Period(PeriodMixin):
     def __init__(
         self,
-        value: Any = ...,
-        freqstr: Any = ...,
-        ordinal: Any = ...,
-        year: Any = ...,
-        month: int = ...,
-        quarter: Any = ...,
-        day: int = ...,
-        hour: int = ...,
-        minute: int = ...,
-        second: int = ...,
+        value: Period | str | None = ...,
+        freq: str | BaseOffset | None = ...,
+        ordinal: int | None = ...,
+        year: int | None = ...,
+        month: int | None = ...,
+        quarter: int | None = ...,
+        day: int | None = ...,
+        hour: int | None = ...,
+        minute: int | None = ...,
+        second: int | None = ...,
     ) -> None: ...
-    def __add__(self, other) -> Period: ...
-    def __eq__(self, other) -> bool: ...
-    def __ge__(self, other) -> bool: ...
-    def __gt__(self, other) -> bool: ...
+    @overload
+    def __sub__(self, other: _PeriodAddSub) -> Period: ...
+    @overload
+    def __sub__(self, other: Period) -> BaseOffset: ...
+    @overload
+    def __sub__(self, other: NaTType) -> NaTType: ...
+    @overload
+    def __sub__(self, other: PeriodIndex) -> Index: ...
+    @overload
+    def __sub__(self, other: TimedeltaSeries) -> PeriodSeries: ...
+    @overload
+    def __sub__(self, other: TimedeltaIndex) -> PeriodIndex: ...
+    def __rsub__(self, other: PeriodIndex) -> Index: ...
+    @overload
+    def __add__(self, other: _PeriodAddSub) -> Period: ...
+    @overload
+    def __add__(self, other: NaTType) -> NaTType: ...
+    @overload
+    def __add__(self, other: Index) -> PeriodIndex: ...
+    @overload
+    def __add__(self, other: TimedeltaSeries) -> PeriodSeries: ...
+    @overload  # type: ignore[override]
+    def __eq__(self, other: Period) -> bool: ...
```

Comment:
Doesn't it override the default so `Period()== object()` isn't allowed except for listed types?  I thought it made sense to do this since the return type of generic objects() isn't always a `bool` - it can be a Series, Index, ndarray.

Reply:
There are two schools of thought here, and, tbh, I'm not sure which is best.  

1. Looking at typeshed, the huge majority of the libraries define `__eq__` using an `object` type for the `other` argument (if they define it at all).  That's because `__eq__()` should accept any object.  In the few cases where they specify another type, it is similar to how we need a special case for `def __eq__(self, other: PeriodIndex) -> npt.NDArray[np.bool_]: ...` but don't define it both ways.  So my suggestion here is that we just include `__eq__()` and `__ne__()` for when `other` is of type `PeriodIndex`.  For example, the following code is valid:  `pd.Period("4Q2005") == 10` so that would mean the type checker would accept it, even though we know the answer is `False`.
2. If we go your direction, we are saying that users should not do equality or inequality comparisons to an object of the wrong type.  But it is valid python.  So maybe we should consider something like this (although I'm not sure it would work without `# type: ignore` annotations:
```python
    @overload  # type: ignore[override]
    def __eq__(self, other: Period) -> bool: ...
    @overload
    def __eq__(self, other: PeriodIndex) -> npt.NDArray[np.bool_]: ...
    @overload
    def __eq__(self, other: object) -> Literal[False]: ...
```

Using something like the second idea is a compromise (if you can get it to work), but I'm leaning towards the first way because it seems to be what others are doing.

Interestingly, the following happens with numpy:
```python
import numpy as np
from numpy import typing as npt
a: npt.NDArray[np.int64] = np.array([1,2,3])
reveal_type(a)
reveal_type(a == 4)
reveal_type(a >= 4)
```

With mypy, you get:
```text
goo.py:4: note: Revealed type is "numpy.ndarray[Any, numpy.dtype[numpy.signedinteger[numpy._typing._64Bit]]]"
goo.py:5: note: Revealed type is "Any"
goo.py:6: note: Revealed type is "numpy.ndarray[Any, numpy.dtype[numpy.bool_]]"
```

So for the `>=` operator, it gives the right type, but for the equality operator, it doesn't.

Reply:
I've gone down the router for `Literal[False]` for `__eq__(object)`